File: src/test/java/org/openrewrite/kotlin/format/SpacesTest.java
/*
 * Copyright 2023 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.kotlin.format;

import org.junit.jupiter.api.Test;
import org.openrewrite.ExecutionContext;
import org.openrewrite.Issue;
import org.openrewrite.internal.ListUtils;
import org.openrewrite.java.tree.Space;
import org.openrewrite.java.tree.TextComment;
import org.openrewrite.kotlin.KotlinIsoVisitor;
import org.openrewrite.kotlin.tree.KSpace;
import org.openrewrite.marker.Markers;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.kotlin.Assertions.kotlin;
import static org.openrewrite.test.RewriteTest.toRecipe;

class SpacesTest implements RewriteTest {

    // This test ensures spaces stored in markers are visited and may be removed once the SpacesVisitor and tests are implemented.
    @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/192")
    @SuppressWarnings("RedundantNullableReturnType")
    @Test
    void visitsMarkerLocation() {
        rewriteRun(
          spec -> spec.recipe(toRecipe(() -> new KotlinIsoVisitor<>() {
              @Override
              public Space visitSpace(Space space, KSpace.Location loc, ExecutionContext executionContext) {
                  if (!space.getComments().isEmpty()) {
                      return space;
                  }
                  if (loc == KSpace.Location.TYPE_REFERENCE_PREFIX || loc == KSpace.Location.IS_NULLABLE_PREFIX || loc == KSpace.Location.CHECK_NOT_NULL_PREFIX) {
                      return space.withComments(ListUtils.concat(new TextComment(true, loc.name(), "", Markers.EMPTY), space.getComments()));
                  }
                  return super.visitSpace(space, loc, executionContext);
              }
          })),
          kotlin(
            """
              class A {
                  fun method ( ) : Int ? {
                      return 1
                  }
              }
              """,
            """
              class A {
                  fun method ( ) /*TYPE_REFERENCE_PREFIX*/: Int /*IS_NULLABLE_PREFIX*/? {
                      return 1
                  }
              }
              """
          ),
          kotlin(
            """
              val a = A ( )
              val b = a . method ( ) !!
              val c = b !!
              """,
            """
              val a = A ( )
              val b = a . method ( ) /*CHECK_NOT_NULL_PREFIX*/!!
              val c = b /*CHECK_NOT_NULL_PREFIX*/!!
              """
          )
        );
    }
}

File: rewrite-go/pkg/format/spaces_test.go
/*
 * Copyright 2026 the original author or authors.
 *
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://docs.moderne.io/licensing/moderne-source-available-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package format

import (
	"testing"

	"github.com/openrewrite/rewrite/rewrite-go/pkg/parser"
)

func TestUnaryOperandSpacing(t *testing.T) {
	sources := []string{
		"package p\n\nvar A = + +0\n",
		"package p\n\nvar A = - -0\n",
		"package p\n\nvar A = ^ ^0\n",
		"package p\n\nvar A = !  !true\n",
		"package p\n\nvar A = -  0\n",
		"package p\n\nvar A = 1 - -0\n",
	}
	for _, src := range sources {
		t.Run(src, func(t *testing.T) {
			p := parser.NewGoParser()
			p.ParseOnly = true
			cu, err := p.Parse("t.go", src)
			if err != nil {
				t.Fatalf("parse: %v", err)
			}
			want, err := gofmtSource("t.go", src)
			if err != nil {
				t.Fatalf("gofmt: %v", err)
			}
			if got := runAutoFormat(cu); got != want {
				t.Errorf("got  %q\nwant %q", got, want)
			}
		})
	}
}
